### ISSUE TITLE
All primitives can be spread in objects according to MDN but TypeScript prevents it

### ISSUE THREAD

**Issue by geryogam** [microsoft/TypeScript#57957]
### 🔎 Search Terms

"spreading primitives in objects"

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about spread.

### ⏯ Playground Link

https://www.typescriptlang.org/play?#code/MYewdgzgLgBAHjAvDA3jAdJgjAGg5qAJwFcBTPTdAMwEMAbCc-dMYuuizYsAE1KoCWYUj07oA5OJgBfANxA

### 💻 Code

```ts
const x = { ...1, ...true, ...false, ...null, ...undefined, ...'' };
```

### 🙁 Actual behavior

I got this TypeScript error:

> error TS2698: Spread types may only be created from object types.

It is wrong because of this quote from MDN on [Spread Syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax#description):

> All [primitives](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#primitive_values) can be spread in objects. Only strings have enumerable own properties, and spreading anything else doesn't create properties on the new object.
> 
> ```js
> const obj = { ...true, ..."test", ...10 };
> // { '0': 't', '1': 'e', '2': 's', '3': 't' }
> ```

### 🙂 Expected behavior

I expected no error and `x` to be equal to `{}`.

**Comment by MartinJohns** [comment]
For `null` and `undefined` it's intentional: #57501. I suspect for the other types as well. Spreading them makes no sense and indicates a logic error.

**Comment by fatcerberus** [comment]
[What JavaScript allows at runtime and what TS allows at compile time are intentionally very different things.](https://stackoverflow.com/questions/41750390/what-does-all-legal-javascript-is-legal-typescript-mean)

Since these are no-ops, there's no reason to ever write e.g. `{ ...true }` literally instead of just `{}`.  And if you have something like `boolean | Foo` where it's only *sometimes* an object that you're spreading in, then that's likely to indicate a bug in your code, so TS errors accordingly.

**Comment by nmain** [comment]
Related https://github.com/microsoft/TypeScript/issues/38469

**Comment by geryogam** [comment]
Thanks for the quick feedback.

The use case is this (we have a lot of instances of it in the codebase at my current company):

```ts
const someObject = {
  someProperty,
  ...(condition && { someConditionalProperty })
};
```

It is perfectly valid in JavaScript but in TypeScript we have to write this instead, which is a little awkward:

```ts
const someObject = {
  someProperty,
  ...(condition ? { someConditionalProperty } : {})
};
```

**Comment by RyanCavanaugh** [comment]
This checks without error:
```ts
declare const someProperty: { x: string };
declare const someCondition: boolean;
declare const someConditionalProperty: { y: string };
const someObject = {
  someProperty,
  ...(someCondition && { someConditionalProperty })
};
```
What are your actual types?

**Comment by geryogam** [comment]
The inferred type of `someCondition` is `string | null`:

```ts
import { useSelector } from 'react-redux';

const someCondition = useSelector(state => state.someCondition);
```

What is weird is that the spread does type check without error if I use a type assertion:

```ts
import { useSelector } from 'react-redux';

const someCondition = useSelector(state => state.someCondition) as string | null;
```

**Comment by MartinJohns** [comment]
> The inferred type of `someCondition` is `string | null`:

Are you absolutely sure? Because then it should work. That it does work when you use a type assertion indicates the type is not as you think it is.

**Comment by RyanCavanaugh** [comment]
You seem to be reducing these examples so far that they stop working or making sense. I'd really like an actual example of this problem occurring to make sure nothing weird is happening.

**Comment by geryogam** [comment]
> Are you absolutely sure? Because then it should work. That it does work when you use a type assertion indicates the type is not as you think it is.

`string | null` is the inferred type that VSCode showed me. But after rebasing my branch, TypeScript does not complain anymore, I can’t reproduce the error. I don’t know what happened but that’s good news.

I’ve finally understood TypeScript’s behaviour: you can actually spread any primitives, including `null` and `undefined`, since it is allowed in JavaScript, but there must be a case where you spread an `object`:

```ts
const x = new Date().getSeconds() > 30 ? true : null;
const y = { ...(x && {}) };  // okay
```

```ts
const x = new Date().getSeconds() > 30 ? false : null;
const y = { ...(x && {}) };  // error
```

```ts
const x = new Date().getSeconds() > 30 ? true : undefined;
const y = { ...(x && {}) };  // okay
```

```ts
const x = new Date().getSeconds() > 30 ? false : undefined;
const y = { ...(x && {}) };  // error
```

```ts
const x = new Date().getSeconds() > 30 ? true : 0;
const y = { ...(x && {}) };  // okay
```

```ts
const x = new Date().getSeconds() > 30 ? false : 0;
const y = { ...(x && {}) };  // error
```

```ts
const x = new Date().getSeconds() > 30 ? true : '';
const y = { ...(x && {}) };  // okay
```

```ts
const x = new Date().getSeconds() > 30 ? false : '';
const y = { ...(x && {}) };  // error
```

So thank you all for your helpful comments. I’m closing this issue now.

**Comment by MartinJohns** [comment]
> I’ve finally understood TypeScript’s behaviour: you can actually spread any primitives, including `null` and `undefined`, since it’s allowed in JavaScript, but there must be a case where you spread an `object`:

Yes, as mentioned in this comment: https://github.com/microsoft/TypeScript/issues/20153#issuecomment-516193778

If you try to spread something that's known to **always** be a primitive (except string), then it indicates an error in your logic. Why would you want to spread that value? But if it **may** be an object (like in a union) it's allowed.

**Comment by geryogam** [comment]
Exactly. Could we improve the error message so that developers are not confused like I was?

Instead of this message which is too restrictive

> error TS2698: Spread types may only be created from object types.

what do you think of this message?

> error TS2698: Spread types may only be created from types that include object types.

Is it the standard wording in type theory to express the idea?

**Comment by RyanCavanaugh** [comment]
For things that are harmless but appear wrong, we've had good feedback (i.e. lack of complaint) about messages like "Spread appears unintentional because this value is not possibly an object". Open to discussion.

**Comment by geryogam** [comment]
I like this wording too. To proceed forward, should I reopen this issue, create another issue, or do nothing?

**Comment by RyanCavanaugh** [comment]
Just a PR is fine

**Comment by geryogam** [comment]
Done.
